README.md
| Appveyor | Appveyor Tests  | PS Gallery Downloads | PS Gallery Version|
|--------|--------|--------|--------|
[![AppVeyor][appveyor-badge]][appveyor] | [![AppVeyor][appveyor-badge-tests]][appveyor] | [![PowerShell Gallery][powershellgallery-downloads]][powershellgallery] | [![PowerShell Gallery][powershellgallery-version]][powershellgallery]

# PSPuppetOrchestrator

PowerShell module for calling the Puppet Orchestrator API.

## Installation

```powershell
PS> Install-Module -Name PSPuppetOrchestrator
```

## Included Functions

`Invoke-PuppetTask`

`Get-PuppetTask`

`Get-PuppetTasks`

`Get-PuppetJob`

`Get-PuppetJobReport`

`Get-PuppetJobResults`

`Get-PuppetPCPNodeBrokerDetails`

`Get-PuppetAuthToken`

## Token

In order to call the `orchestrator` API endpoint you'll need a user token. You can call the `Get-PuppetAuthToken` cmdlet to acquire a token. You can use the `-SkipCertificateCheck` switch in case your Puppet Master Console certificate has not been imported into your work station certificate store.

If you want to use this cmdlet in an automated environment make sure you create the credential object in your script. If you do not pass a credential the cmdlet will hang the script waiting for you to enter credentials at the commandline.

```powershell
$creds = Get-Credential

$token = Get-PuppetAuthToken -Master '<Master FQDN>' -SkipCertificateCheck -Credential $creds
```

OUTPUT

```plaintext
***********************************wCB8
```

## Examples

### Get tasks

Get a list of tasks.

```powershell
PS> Get-PuppetTasks -token $token -master $master
```

OUTPUT

```plaintext
id                                                                       name                            permitted
--                                                                       ----                            ---------
https://puppet:8143/orchestrator/v1/tasks/powershell_tasks/getkb         powershell_tasks::getkb         True
https://puppet:8143/orchestrator/v1/tasks/powershell_tasks/account_audit powershell_tasks::account_audit True
https://puppet:8143/orchestrator/v1/tasks/powershell_tasks/switch        powershell_tasks::switch        True
https://puppet:8143/orchestrator/v1/tasks/powershell_tasks/ps1exec       powershell_tasks::ps1exec       True
https://puppet:8143/orchestrator/v1/tasks/powershell_tasks/disablesmbv1  powershell_tasks::disablesmbv1  True
```

### Run tasks

`Invoke-PuppetTask` can target nodes three different ways by specifying **one** of the following parameters.

- Nodes
  - EXAMPLE: `-Nodes @('node1.contoso.com','node2.contoso.com','node3.contoso.com','node4.contoso.com')`
- Query
  - EXAMPLE: `-Query '["from", "inventory", ["=", "facts.os.name", "windows"]]'`
- Node_group
  - EXAMPLE: `-Node_group '7a692b61-8087-4452-9cf8-58ed2acee2a0'`

#### Invoke-PuppetTask Parameter Splat Examples

Create an `Invoke-PuppetTask` parameter splat to target and array of nodes.

```powershell
$invokePuppetTaskSplat = @{
    Token  = $token
    Master = $master
    Task   = 'jpi::get_dns'
    Nodes  = @('node1.contoso.com','node2.contoso.com','node3.contoso.com','node4.contoso.com')
}
```

Create an `Invoke-PuppetTask` parameter splat to target nodes from a PQL query.

```powershell
$invokePuppetTaskSplat = @{
    Token  = $token
    Master = $master
    Task   = 'jpi::get_dns'
    Query  = '["from", "inventory", ["=", "facts.os.name", "windows"]]'
}
```

Create an `Invoke-PuppetTask` parameter splat to target a node group.

```powershell
$invokePuppetTaskSplat = @{
    Token      = $token
    Master     = $master
    Task       = 'jpi::get_dns'
    Node_group = '7a692b61-8087-4452-9cf8-58ed2acee2a0'
}
```

Create an `Invoke-PuppetTask` parameter splat specifying additional parameters for the `jpi::set_dns` Puppet Task.

```powershell
$invokePuppetTaskSplat = @{
    Token            = $token
    Master           = $master
    Task             = 'jpi::set_dns'
    Parameters       = @{
        primary   = '10.0.3.24'
        secondary = '10.0.3.22'
        tertiary  = '8.8.8.8'
    }
    Nodes            = @('node1.contoso.com', 'node2.contoso.com')
}
```

Create an `Invoke-PuppetTask` parameter splat specifying additional parameters for the `jpi::set_dns` Puppet Task and `-Wait` 120 seconds for the job to complete. **Note, the job will continue to run even if the job exceeds the `-Wait` time of 120 seconds.**

```powershell
$invokePuppetTaskSplat = @{
    Token            = $token
    Master           = $master
    Task             = 'jpi::set_dns'
    Parameters       = @{
        primary   = '10.0.3.24'
        secondary = '10.0.3.22'
        tertiary  = '8.8.8.8'
    }
    Nodes            = @('node1.contoso.com', 'node2.contoso.com')
    Wait             = 120
}
```

#### Invoke-PuppetTask Outputs

When `Invoke-PuppetTask` is supplied the `-Wait` parameter and the job completes within the wait time you'll be returned the full job results (identical to the results given by `Get-PuppetJob`).

OUTPUT

```plaintext
description :
report      : @{id=https://puppet.contoso.us:8143/orchestrator/v1/jobs/1326/report}
name        : 1326
events      : @{id=https://puppet.contoso.us:8143/orchestrator/v1/jobs/1326/events}
command     : task
type        : task
state       : finished
nodes       : @{id=https://puppet.contoso.us:8143/orchestrator/v1/jobs/1326/nodes}
status      : {@{state=ready; enter_time=4/8/2020 6:22:21 PM; exit_time=4/8/2020 6:22:21 PM}, @{state=running; enter_time=4/8/2020 6:22:21 PM; exit_time=4/8/2020 6:22:24 PM}, @{state=finished; enter_time=4/8/2020 6:22:24 PM; exit_time=}}
id          : https://puppet.contoso.us:8143/orchestrator/v1/jobs/1326
environment : @{name=production}
options     : @{description=; transport=pxp; noop=False; task=jpi::set_dns; sensitive=System.Object[]; scheduled-job-id=; params=; scope=; environment=production}
timestamp   : 4/8/2020 6:22:24 PM
owner       : @{email=; is_revoked=False; last_login=4/8/2020 6:17:18 PM; is_remote=False; login=admin; is_superuser=True; id=42bf351c-f9ec-40af-84ad-e976fec7f4bd; role_ids=System.Object[]; display_name=Administrator; is_group=False}
node_count  : 1
node_states : @{finished=1}
```

When `Invoke-PuppetTask` is not supplied the `-Wait` parameter you'll immediately be returned the `id` and `name` of the job.

OUTPUT

```plaintext

id                                                       name
--                                                       ----
https://puppet.piccola.us:8143/orchestrator/v1/jobs/1324 1324
```

### Get task report

Using the `name` value of `1324` returned from previously executed `Invoke-PuppetTask` get the report.

```powershell
PS> Get-PuppetJobReport -Master $master -Token $token -ID 1324
```

OUTPUT

```plaintext
node             : node1.contoso.com
state            : finished
start_timestamp  : 4/8/2020 4:45:06 PM
finish_timestamp : 4/8/2020 4:45:08 PM
timestamp        : 4/8/2020 4:45:08 PM
events           : {}
```

### Get tasks results

Using the `name` returned from `Invoke-PuppetTask` get the results. Depending on the depth of the structured data returned from the task you may have to "dot explore" into it.

```powershell
PS> Get-PuppetJobResults -Master $master -Token $token -ID 1324 -OutVariable 'PuppetJobResults'
```

OUTPUT

```plaintext
PS> $PuppetJobResults
    transport        : pcp
    finish_timestamp : 4/8/2020 4:45:08 PM
    transaction_uuid :
    start_timestamp  : 4/8/2020 4:45:06 PM
    name             : node1.contoso.com
    duration         : 2.275
    state            : finished
    details          :
    result           : @{after=; before=}
    latest-event-id  : 9510
    timestamp        : 4/8/2020 4:45:08 PM
```

Though very specific to the `jpi::set_dns` Puppet Task, below is the extended result data returned from job `1324`. Retrieved with `$PuppetJobResults.result | ConvertTo-Json`.

```json
{
  "after": {
    "Index": 4,
    "IPAddress": [
      "10.0.3.117",
      "fe80::987b:9b0a:1532:8d2d"
    ],
    "MACAddress": "00:50:56:89:53:29",
    "DNSServerSearchOrder": [
      "10.0.3.24",
      "10.0.3.22",
      "8.8.8.8"
    ]
  },
  "before": {
    "Index": 4,
    "IPAddress": [
      "10.0.3.117",
      "fe80::987b:9b0a:1532:8d2d"
    ],
    "MACAddress": "00:50:56:89:53:29",
    "DNSServerSearchOrder": [
      "10.0.3.24",
      "10.0.3.22",
      "8.8.8.8"
    ]
  }
}
```

## License

PSPuppetOrchestrator is released under the [MIT license](http://www.opensource.org/licenses/MIT).

[appveyor]: https://ci.appveyor.com/project/joeypiccola/pspuppetorchestrator
[appveyor-badge]: https://ci.appveyor.com/api/projects/status/6g7fk7xes4vn5fog/branch/master?svg=true&passingText=master%20-%20PASSING&pendingText=master%20-%20PENDING&failingText=master%20-%20FAILING
[appveyor-badge-tests]: https://img.shields.io/appveyor/tests/joeypiccola/pspuppetorchestrator/master
[powershellgallery]: https://www.powershellgallery.com/packages/PSPuppetOrchestrator
[powershellgallery-downloads]: https://img.shields.io/powershellgallery/dt/pspuppetorchestrator
[powershellgallery-version]: https://img.shields.io/powershellgallery/v/pspuppetorchestrator


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.1.9]
### Added
- Move to 100% PowerShellBuild `0.4.0` build process.

## [0.1.8]
### Added
- `Get-PuppetAuthToken` You can now provide a server name for your Puppet master and optionally a listening port, along with a credential, and easily obtain an api auth token for use with the rest of the cmdlets in the module. (via [RandomNoun7](https://github.com/RandomNoun7))

## [0.1.7]
### Changed
- `Invoke-PuppetTask` can now target `nodes`, `query`, or `node_group`. This parameters on `Invoke-PuppetTask` have been changed and are not backwards compatible with version 0.1.6 and below.

## [0.1.6]
### Added
- added missing pester tests although skipped at the moment

## [0.1.5]
### Added
- added pester tests
- token nodes in readme
- be consistent with how we write out object with > item

## [0.1.4]
### Added
- bump build reqs
- changed md docs location
- added function docs
- updated readme

## [0.1.3]
### Added
- Adding more docs.

## [0.1.2]
### Added
- Adding more docs.

## [0.1.1]
### Added
- Adding docs.

## [0.1.0]
### Added
- Initial release.


puppet_API.ps1
Function Get-PuppetJobReport {
    Param(
        [Parameter(Mandatory)]
        [int]$ID,
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master
    )

    $hoststr = "https://$master`:8143/orchestrator/v1/jobs/$id/report"
    $headers = @{'X-Authentication' = $Token}

    $result  = Invoke-WebRequest -Uri $hoststr -Method Get -Headers $headers
    $content = $result.content | ConvertFrom-Json
    $report  = $content.report

    foreach ($server in $report) {
        write-output $server
    }
}

Function Get-PuppetJob {
    Param(
        [Parameter(Mandatory)]
        [int]$ID,
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master
    )

    $hoststr = "https://$master`:8143/orchestrator/v1/jobs/$id"
    $headers = @{'X-Authentication' = $Token}

    $result  = Invoke-WebRequest -Uri $hoststr -Method Get -Headers $headers
    $content = $result.content | ConvertFrom-Json

    Write-Output $content
}

Function Get-PuppetTasks {
    Param(
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master
    )

    $hoststr = "https://$master`:8143/orchestrator/v1/tasks"
    $headers = @{'X-Authentication' = $Token}

    $result  = Invoke-WebRequest -Uri $hoststr -Method Get -Headers $headers
    $content = $result | ConvertFrom-Json

    Write-Output $content.items
}

Function Get-PuppetTask {
    Param(
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master,
        [Parameter()]
        [string]$Module,
        [Parameter(Mandatory)]
        [string]$Name
    )

    $hoststr = "https://$master`:8143/orchestrator/v1/tasks/$Module/$Name"
    $headers = @{'X-Authentication' = $Token}

    # try and get the task in it's standard form $moduleName/$taskName
    try {
        $result  = Invoke-WebRequest -Uri $hoststr -Method Get -Headers $headers -ErrorAction SilentlyContinue
    } catch {
        # try and get the task again assuming it's built in with a default task name of 'init' (e.g. reboot/init)
        try {
            $hoststr = "https://$master`:8143/orchestrator/v1/tasks/$Module/init"
            $result  = Invoke-WebRequest -Uri $hoststr -Method Get -Headers $headers
        } catch {
            Write-Error $_.exception.message
        }
    }

    if ($result) {
        $content = $result | ConvertFrom-Json
        Write-Output $content
    }
}

Function Get-PuppetJobNodes {
    Param(
        [Parameter(Mandatory)]
        [int]$ID,
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master
    )

    $hoststr = "https://$master`:8143/orchestrator/v1/jobs/$id/nodes"
    $headers = @{'X-Authentication' = $Token}

    $result  = Invoke-WebRequest -Uri $hoststr -Method Get -Headers $headers
    $content = $result.content | ConvertFrom-Json

    Write-Output $content.items
}

Function Invoke-PuppetTask {
<#
$scope = @('den3-node-1.ad.piccola.us','den3-node-3.ad.piccola.us','den3-node-4.ad.piccola.us','den3-node-5.ad.piccola.us')

$splat = @{
    Token = 'AKSeiFY0KJkQxUe6ovYfRQ9JbKjXY8So1GoCApkAMzOZ'
    Master = 'puppet.piccola.us'
    Task = 'powershell_tasks::disablesmbv1'
    Environment = 'development'
    Parameters = [PSCustomObject]@{
        action = 'get'
    }
    Description = 'This was run from PowerShell'
    Scope = $scope
    ScopeType = 'nodes'
}

Invoke-PuppetTask @splat -Wait -Timeout 120
#>
    Param(
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master,
        [Parameter(Mandatory)]
        [string]$Task,
        [Parameter()]
        [string]$Environment = 'production',
        [Parameter()]
        [PSCustomObject]$Parameters = @{},
        [Parameter()]
        [string]$Description = '',
        [Parameter(Mandatory)]
        [PSCustomObject[]]$Scope,
        [Parameter(Mandatory)]
        [ValidateSet('nodes')]
        [string]$ScopeType,
        [Parameter()]
        [switch]$Wait,
        [Parameter()]
        [int]$Timeout = 300
    )

    $req = [PSCustomObject]@{
        environment = $Environment
        task        = $Task
        params      = $Parameters
        description = $Description
        scope       = [PSCustomObject]@{
        $ScopeType = $Scope
        }
    } | ConvertTo-Json

    $hoststr = "https://$master`:8143/orchestrator/v1/command/task"
    $headers = @{'X-Authentication' = $Token}

    $result  = Invoke-WebRequest -Uri $hoststr -Method Post -Headers $headers -Body $req
    $content = $result.content | ConvertFrom-Json

    if ($wait) {
        # sleep 5s for the job to register
        Start-Sleep -Seconds 5

        $jobSplat = @{
            token = $Token
            master = $master
            id = $content.job.name
        }

        # create a timespan
        $timespan = New-TimeSpan -Seconds $timeout
        # start a timer
        $stopwatch = [diagnostics.stopwatch]::StartNew()

        # get the job state every 5 seconds until our timeout is met
        while ($stopwatch.elapsed -lt $timespan) {
            # optoins are new, ready, running, stopping, stopped, finished, or failed
            $job = Get-PuppetJob @jobSplat
            if (($job.State -eq 'stopped') -or ($job.State -eq 'finished') -or ($job.State -eq 'failed')) {
                $taskJobContent = [PSCustomObject]@{
                    task = $content
                    job = $job
                }
                Write-Output $taskJobContent
                break
            }
            Start-Sleep -Seconds 5
        }
        if ($stopwatch.elapsed -ge $timespan) {
            Write-Error "Timeout of $Timeout`s has exceeded."
            break
        }
    } else {
        Write-Output $content
    }
}

function Get-PuppetPCPNodeBrokerDetails {
    Param(
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master,
        [Parameter(Mandatory)]
        [string]$Node
    )

    $hoststr = "https://$master`:8143/orchestrator/v1/inventory/$node"
    $headers = @{'X-Authentication' = $Token}

    $result  = Invoke-WebRequest -Uri $hoststr -Method Get -Headers $headers
    $content = $result | ConvertFrom-Json
    Write-Output $content
}

Function Wait-PuppetNodePCPBroker {
    Param(
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master,
        [Parameter(Mandatory)]
        [string]$Node,
        [Parameter()]
        [int]$Timeout = 300
    )

    $detailsSplat = @{
        token = $Token
        master = $master
        node = $node
    }

    # create a timespan
    $timespan = New-TimeSpan -Seconds $timeout
    # start a timer
    $stopwatch = [diagnostics.stopwatch]::StartNew()

    # get the broker status every 5 seconds until our timeout is met
    while ($stopwatch.elapsed -lt $timespan) {
        # get the broker status
        if (($one = Get-PuppetPCPNodeBrokerDetails @detailsSplat).connected -eq $false) {
            # broker status is disconnected, sleep 5s and check agian to confirm not a blip or false positive
            Write-Verbose "Broker status is $($one.connected), (timeout: $($stopwatch.elapsed.TotalSeconds)s of $Timeout`s elapsed)."
            Write-Verbose "Sleping 5 seconds and checking again."
            Start-Sleep -Seconds 5
            if (($two = Get-PuppetPCPNodeBrokerDetails @detailsSplat).connected -eq $false) {
                Write-Verbose "Broker status is still $($two.connected), (timeout: $($stopwatch.elapsed.TotalSeconds)s of $Timeout`s elapsed)."
                # broker status is disconnected, break out of the loop
                break
            }
        } else {
            Write-Verbose "Broker status is $($one.connected), (timeout: $($stopwatch.elapsed.TotalSeconds)s of $Timeout`s elapsed)."
        }
        Start-Sleep -Seconds 5
    }
    if ($stopwatch.elapsed -ge $timespan) {
        Write-Error "Timeout of $Timeout`s has exceeded."
        break
    }

    Write-Verbose "$Node broker status confirmed disconnected."

    # get the broker status every 5 seconds until our timeout is met
    while ($stopwatch.elapsed -lt $timespan) {
        # get the broker status
        if (($three = Get-PuppetPCPNodeBrokerDetails @detailsSplat).connected -eq $true) {
            # broker status is connected, sleep 5s and check agian to confirm not a blip or false positive
            Write-Verbose "Broker status is $($three.connected), (timeout: $($stopwatch.elapsed.TotalSeconds)s of $Timeout`s elapsed)."
            Write-Verbose "Sleping 5 seconds and checking again."
            Start-Sleep -Seconds 5
            if (($four = Get-PuppetPCPNodeBrokerDetails @detailsSplat).connected -eq $true) {
                Write-Verbose "Broker status is still $($four.connected), (timeout: $($stopwatch.elapsed.TotalSeconds)s of $Timeout`s elapsed)."
                # broker status is connected, break out of the loop
                break
            }
        } else {
            Write-Verbose "Broker status is $($three.connected), (timeout: $($stopwatch.elapsed.TotalSeconds)s of $Timeout`s elapsed)."
        }
        Start-Sleep -Seconds 5
    }
    if ($stopwatch.elapsed -ge $timespan) {
        Write-Error "Timeout of $Timeout`s has exceeded."
        break
    }

    Write-Verbose "$Node broker status confirmed connected."
}

Function Get-PuppetNodeFacts {
    Param(
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master,
        [Parameter(Mandatory)]
        [string]$Node
    )

    $hoststr = "https://$master`:8081/pdb/query/v4/facts"
    $headers = @{'X-Authentication' = $Token}

    $query = @{
        query = "[""="", ""certname"", ""$node""]"
    }

    $result = Invoke-WebRequest -Uri $hoststr -Method Get -Body $query -Headers $headers
    $content = $result | ConvertFrom-Json

    # borrowed from Warren Frame
    $h = @{}
    $content | foreach-object { [void]$h.set_item($_.Name, $_.Value) }
    [pscustomobject]$h

}

Function Get-PuppetNodeFact {
    Param(
        [Parameter(Mandatory)]
        [string]$Token,
        [Parameter(Mandatory)]
        [string]$Master,
        [Parameter(Mandatory)]
        [string]$Node,
        [Parameter(Mandatory)]
        [string]$Fact
    )

    # This is a shortcut to the /pdb/query/v4/facts endpoint.
    $hoststr = "https://$master`:8081/pdb/query/v4/nodes/$node/facts/$Fact"
    $headers = @{'X-Authentication' = $Token}

    $result = Invoke-WebRequest -Uri $hoststr -Method Get -Headers $headers
    $content = $result.content | ConvertFrom-Json

    Write-Output $content.value
}